aiida/tools/importexport/common/zip_folder.py

```diff
@@ -0,0 +1,152 @@
+        if self._buffer is not None:
+            raise IOError('Cannot open again!')
+        self._buffer = io.StringIO()
+
+    def write(self, data):
+        self._buffer.write(data)
+
+    def close(self):
+        self._buffer.seek(0)
+        self._zipfile.writestr(self._fname, self._buffer.read())
+        self._buffer = None
+
+    def __enter__(self):
+        self.open()
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+
+class ZipFolder:
+    """
+    To improve: if zipfile is closed, do something
+    (e.g. add explicit open method, rename open to openfile,
+    set _zipfile to None, ...)
+    """
+
+    def __init__(self, zipfolder_or_fname, mode=None, subfolder='.', use_compression=True, allowZip64=True):
+        """
+        :param zipfolder_or_fname: either another ZipFolder instance,
+          of which you want to get a subfolder, or a filename to create.
+        :param mode: the file mode; see the zipfile.ZipFile docs for valid
+          strings. Note: can be specified only if zipfolder_or_fname is a
+          string (the filename to generate)
+        :param subfolder: the subfolder that specified the "current working
+          directory" in the zip file. If zipfolder_or_fname is a ZipFolder,
+          subfolder is a relative path from zipfolder_or_fname.subfolder
+        :param use_compression: either True, to compress files in the Zip, or
+          False if you just want to pack them together without compressing.
+          It is ignored if zipfolder_or_fname is a ZipFolder isntance.
+        """
+        if isinstance(zipfolder_or_fname, str):
+            the_mode = mode
+            if the_mode is None:
+                the_mode = 'r'
+            if use_compression:
+                compression = zipfile.ZIP_DEFLATED
+            else:
+                compression = zipfile.ZIP_STORED
+            self._zipfile = zipfile.ZipFile(
+                zipfolder_or_fname, mode=the_mode, compression=compression, allowZip64=allowZip64
+            )
+            self._pwd = subfolder
+        else:
+            if mode is not None:
+                raise ValueError("Cannot specify 'mode' when passing a ZipFolder")
+            self._zipfile = zipfolder_or_fname._zipfile  # pylint: disable=protected-access
+            self._pwd = os.path.join(zipfolder_or_fname.pwd, subfolder)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    def close(self):
+        self._zipfile.close()
+
+    @property
+    def pwd(self):
+        return self._pwd
+
+    def open(self, fname, mode='r'):
+        if mode == 'w':
+            return MyWritingZipFile(zip_file=self._zipfile, fname=self._get_internal_path(fname))
+        # else
+        return self._zipfile.open(self._get_internal_path(fname), mode)
+
+    def _get_internal_path(self, filename):
+        return os.path.normpath(os.path.join(self.pwd, filename))
```

Comment:
shouldn't we be using pathlib?
(several other uses of os.path in this class)

Reply:
Err, I'm going to be lazy here and say not now. I have anyhow allowed for the filename in the __init__ to be a Path and thats the main one